ATO-579: create identity progress lambda

isaac-GDS opened the pull request:
Identity Progress lambda and exposed on frontend api

hannah-leong commented on ci/terraform/oidc/api-gateway-frontend.tf:
```diff
@@ -86,6 +86,7 @@ resource "aws_api_gateway_deployment" "frontend_deployment" {
       local.deploy_check_email_fraud_block_count == 1 ? module.check_email_fraud_block[0].integration_trigger_value : null,
       local.deploy_check_email_fraud_block_count == 1 ? module.check_email_fraud_block[0].method_trigger_value : null,
       local.account_modifiers_encryption_policy_arn,
```
Maybe add commas on the end to match the pattern before

hannah-leong commented on ci/terraform/oidc/api-gateway-frontend.tf:
```diff
@@ -86,6 +86,7 @@ resource "aws_api_gateway_deployment" "frontend_deployment" {
       local.deploy_check_email_fraud_block_count == 1 ? module.check_email_fraud_block[0].integration_trigger_value : null,
       local.deploy_check_email_fraud_block_count == 1 ? module.check_email_fraud_block[0].method_trigger_value : null,
       local.account_modifiers_encryption_policy_arn,
+      var.orch_identity_progress_enabled
```
```suggestion
      var.orch_identity_progress_enabled,
```

hannah-leong commented on ci/terraform/oidc/api-gateway-frontend.tf:
```diff
@@ -110,6 +111,7 @@ resource "aws_api_gateway_deployment" "frontend_deployment" {
     module.doc-app-authorize,
     module.orch_auth_code,
     module.check_reauth_user,
+    aws_api_gateway_integration.orch_identity_progress_integration
```
```suggestion
    aws_api_gateway_integration.orch_identity_progress_integration,
```

hannah-leong commented on ci/terraform/oidc/api-gateway.tf:
```diff
@@ -1283,4 +1282,4 @@ resource "aws_api_gateway_integration" "orch_userinfo_integration" {
   type                    = "AWS_PROXY"
   integration_http_method = "POST"
   uri                     = "arn:aws:apigateway:eu-west-2:lambda:path/2015-03-31/functions/arn:aws:lambda:eu-west-2:${var.orch_account_id}:function:${var.orch_userinfo_name}:latest/invocations"
-}
+}
```
Leave the new line in?

hannah-leong commented on ci/terraform/oidc/api-gateway-frontend.tf:
```diff
@@ -358,3 +360,38 @@ resource "aws_wafv2_web_acl_logging_configuration" "waf_logging_config_frontend_
+  type                    = "AWS_PROXY"
+  integration_http_method = "POST"
+  uri                     = "arn:aws:apigateway:eu-west-2:lambda:path/2015-03-31/functions/arn:aws:lambda:eu-west-2:${var.orch_account_id}:function:${var.orch_identity_progress_name}:latest/invocations"
+}
```
Leave the new line in?

hannah-leong commented on template.yaml:
```diff
@@ -3217,6 +3222,208 @@ Resources:
+              - 'arn:aws:dynamodb:eu-west-2:${AccountId}:table/${AuthEnvironment}-identity-credentials'
+              - AccountId: !FindInMap [ EnvironmentConfiguration, !Ref Environment, authAccountId ]
+                AuthEnvironment: !FindInMap [ EnvironmentConfiguration, !Ref Environment, authEnvironment ]
+          - Sid: AllowDecrpytUserInfoTable
+            Effect: "Allow"
+            Action: "kms:Decrypt*"
+            Resource:
```
For consistency
```suggestion
          - Sid: AllowDecrpytUserInfoTable
            Effect: Allow
            Action: kms:Decrypt*
            Resource:
```

hannah-leong commented on template.yaml:
```diff
@@ -3217,6 +3222,208 @@ Resources:
+            Resource:
+              - !Sub arn:aws:sqs:eu-west-2:${AWS::AccountId}:orchestration-${Environment}-txma-AuditEventQueue
+          - Sid: AllowUserInfoTableReadAccess
+            Effect: "Allow"
```
For consistency:
```suggestion
            Effect: Allow
```

hannah-leong commented on template.yaml:
```diff
@@ -3217,6 +3222,208 @@ Resources:
+            Effect: "Allow"
+            Action: "kms:Decrypt*"
+            Resource:
+              - !FindInMap [ EnvironmentConfiguration, !Ref Environment, userInfoEncryptionKeyArn ]
+          - Sid: AllowGetParameters
+            Effect: "Allow"
+            Action:
+              - "ssm:GetParameter"
+              - "ssm:GetParameters"
+            Resource:
+              - !Sub "arn:aws:ssm:eu-west-2:${AWS::AccountId}:parameter/${Environment}-session-redis-master-host"
+              - !Sub "arn:aws:ssm:eu-west-2:${AWS::AccountId}:parameter/${Environment}-session-redis-tls"
+              - !Sub "arn:aws:ssm:eu-west-2:${AWS::AccountId}:parameter/${Environment}-session-redis-port"
+              - !Sub "arn:aws:ssm:eu-west-2:${AWS::AccountId}:parameter/${Environment}-session-redis-password"
+          - Sid: AllowDecryptOfParameters
+            Effect: "Allow"
+            Action: "kms:Decrypt"
+            Resource: !GetAtt MainKmsKey.Arn
```
More for consistency

hannah-leong reviewed (commented):
Looks good, just some stylistic change requests :)

isaac-GDS commented:
Rebased to main and updated changes but might be worth checking anything i might have missed.

Things updates: 
- Moved away from function name env
- Added checkgov for ATO-657

hannah-leong commented on template.yaml:
```diff
@@ -3339,6 +3344,209 @@ Resources:
+            Action: kms:Decrypt
+            Resource: !GetAtt MainKmsKey.Arn
+      Tags:
+        CheckovRulesToSkip: CKV_AWS_116.CKV_AWS_117.CKV_AWS_173
```
Think 115 should also be here?

isaac-GDS replied:
Good spot!

hannah-leong commented on template.yaml:
```diff
@@ -3339,6 +3344,209 @@ Resources:
+            Action: kms:Decrypt*
+            Resource:
+              - !FindInMap [ EnvironmentConfiguration, !Ref Environment, userInfoEncryptionKeyArn ]
+          - Sid: AllowGetParameters
+            Effect: Allow
+            Action:
+              - "ssm:GetParameter"
+              - "ssm:GetParameters"
+            Resource:
+              - !Sub "arn:aws:ssm:eu-west-2:${AWS::AccountId}:parameter/${Environment}-session-redis-master-host"
+              - !Sub "arn:aws:ssm:eu-west-2:${AWS::AccountId}:parameter/${Environment}-session-redis-tls"
+              - !Sub "arn:aws:ssm:eu-west-2:${AWS::AccountId}:parameter/${Environment}-session-redis-port"
+              - !Sub "arn:aws:ssm:eu-west-2:${AWS::AccountId}:parameter/${Environment}-session-redis-password"
```
Just being picky, but maybe remove quotation marks from these for consistency?

isaac-GDS replied:
Good to be picky :)

hannah-leong reviewed (commented):
Might be worth checking the code scanning comments and dismiss them if they don't apply (haven't looked at them all but looks like some can be dismissed)

hannah-leong commented on template.yaml:
```diff
@@ -3339,6 +3344,209 @@ Resources:
+              - 'arn:aws:dynamodb:eu-west-2:${AccountId}:table/${AuthEnvironment}-identity-credentials'
+              - AccountId: !FindInMap [ EnvironmentConfiguration, !Ref Environment, authAccountId ]
+                AuthEnvironment: !FindInMap [ EnvironmentConfiguration, !Ref Environment, authEnvironment ]
+          - Sid: AllowDecrpytUserInfoTable
```
```suggestion
          - Sid: AllowDecryptUserInfoTable
```

isaac-GDS replied:
Looks like copied over the wrong spelling - good find

isaac-GDS commented:
Rebased - switched userInfoEncryptionKeyArn to authenticationCallbackUserinfoKeyArn

something to look for is if any current policies could be referenced

isaac-GDS commented on template.yaml:
```diff
@@ -2374,6 +2379,149 @@ Resources:
+    Properties:
+      LogGroupName: !Sub /aws/lambda/${Environment}-identity-progress-lambda
+      KmsKeyId: !GetAtt MainKmsKey.Arn
+      RetentionInDays: 14
```
Confirmed that we should be defaulting to 30days. Thanks @kalpaitch for bringing it up. Will update this lambda here and the other lambdas in a separate PR.

ethanmills commented:
@isaac-GDS Shall we close this for now and revisit when the orch frontend is closer?

isaac-GDS commented:
> @isaac-GDS Shall we close this for now and revisit when the orch frontend is closer?

Its ready to be merged - but happy to close to avoid unused code in prod.
